### ISSUE TITLE
Tuist generate with focused target(s) and cached dependencies fails to build due to missing required module  

### ISSUE THREAD

**Issue by KaiOelfke** [tuist/tuist#6485]
### What happened?

Given three targets: App, Framework, NestedFramework that depend on each other like App <- Framework <- NestedFramework and an external dependency (CasePaths in my sample). After using tuist install and tuist cache some tuist generate projects won't compile.

tuist generate (whole project) ✅
tuist generate NestedFramework ✅
tuist generate Framework NestedFramework ✅
tuist generate Framework ❌
tuist generate App Framework ❌
tuist generate App ❌

The build error is: `<unknown>:0: error: missing required module 'CasePaths'`.

I think this is related to:

https://forums.swift.org/t/issue-with-third-party-dependencies-inside-a-xcframework-through-swiftpm/41977/2
https://github.com/tuist/tuist/issues/6430

It is my understanding from investigating the cached xcframework and the swift forum post by NeoNacho that the Swift Compiler doesn't see the required module. 

**Possible solution:**

So I tried to fix it by manually providing a module map. Adding this flag to each generated target depending directly or indirectly on NestedFramework fixes the issue.

Other Swift flags:
`-Xcc`
`-fmodule-map-file=/Users/kaioelfke/.cache/tuist-cloud/BinaryCache/ad5ea32c6188f351c74185ce42b0b321/CasePaths.xcframework/ios-arm64/CasePaths.framework/Modules/module.modulemap`

I can't say if this is the right way or best way to solve this issue. But it seems, if Tuist could auto-generate modulemap files and flags depending on what targets are focused with tuist generate and what dependencies are cached then this should fix it.

I ran into this issue, when converting my biggest app to use Tuist. One of my motivations is more productivity as my app has many targets and binary caching plus focused projects with tuist generate. But due to this issue I can't do this currently.

### How do we reproduce it?

1. Download the sample project
2. Run tuist install
3. Run tuist cache
4. Run the various tuist generate commands from above and try to build

[NoSuchSample.zip](https://github.com/user-attachments/files/16103811/NoSuchSample.zip)

### Error log

```
<unknown>:0: error: missing required module 'CasePaths'
```

### macOS version

14.5

### Tuist version

4.18.0

### Xcode version

15.4

**Comment by KaiOelfke** [comment]
For further clarification there's more complexity here for Swift 6 or with 5.9 and the experimental flag AccessLevelOnImport.

https://github.com/swiftlang/swift-evolution/blob/main/proposals/0409-access-level-on-imports.md

TLDR:

For dependencies that are imported like below it should compile just fine. There's some further edge cases described in the proposal where dependency clients may still need to load the dependency.

```
internal import Dependency
private import Dependency
fileprivate import Dependency
@_implementationOnly import Dependency
```

For  `public import Dependency` we run into this issue. Note that in Swift language mode 5 and 6 by default all `import Dependency` are considered as `public import Dependency`. This only changes in the next language mode after Swift 6 mode. It can be also enabled with `InternalImportsByDefault`.

For `package import Dependency` I'm not sure. I guess within SPM it would be fine, but for packages within Tuist I'm not sure how that behaves.

So maybe Tuist should add some syntax:

```
.external("Dependency", public: true) // Should make sure module maps are added where it's needed or something else for module visibility
.external("Dependency", public: false) // Can exclude dependency as transitive dependency
```

Maybe also with a default value depending on settings and language modes or just the default based on the Xcode version.

**Comment by KaiOelfke** [comment]
Unless the Tuist user overwrites target settings by default the targets generated are not resilient. The build setting for this is BUILD_LIBRARY_FOR_DISTRIBUTION or there's the flags -enable-library-evolution and -emit-module-interface.

https://www.swift.org/blog/library-evolution/

> All dependencies of a non-resilient module must be loaded by transitive clients. This is because types in the module can use types from those dependencies in their storage, and the compiler needs complete information about the storage of non-resilient types in order to emit code correctly. This restriction is discussed further in the Future Directions section.

So even if

```
internal import Dependency
private import Dependency
fileprivate import Dependency
```

is used, but the module is not resilient (default case) then the module map must be added or there's errors. This non-resilient module limitation is something discussed in the future directions section of the proposal.

https://github.com/swiftlang/swift-evolution/blob/main/proposals/0409-access-level-on-imports.md#hiding-dependencies-for-non-resilient-modules

**Comment by fortmarek** [comment]
Hey!

We're looking into this one. This issue is most likely very similar to: https://github.com/tuist/tuist/issues/6514

> So I tried to fix it by manually providing a module map. Adding this flag to each generated target depending directly or indirectly on NestedFramework fixes the issue.

Yes, this is most likely the right solution (or more of a workaround for the limitations of xcframeworks). The tricky part there is to propagate that flag when replacing the source targets with xcframeworks.
